File: Assets/Intelli-verse-X-SDK/MultiplayerKernel/Voice/IVXLiveKitVoiceProvider.cs
// IVXLiveKitVoiceProvider — Unity client implementation of IIVXVoice
// against the LiveKit Unity SDK.
//
// Lifecycle:
//   1. Match session receives a VoiceSessionToken from the kernel.
//   2. Adapter constructs an IVXLiveKitVoiceProvider with the token.
//   3. ConnectAsync() establishes an LK Room with the bearer JWT.
//   4. Publish/subscribe state updates flow over the IVXVoice events.
//
// We deliberately do NOT take a hard dependency on the LiveKit SDK at
// compile time — the IVX SDK ships across many engines and plenty of
// games are voiceless. The LiveKit SDK is loaded via reflection so a
// project without `livekit-unity` can still compile this assembly.
//
// To add the LK SDK to your project:
//   1. UPM: https://github.com/livekit/client-sdk-unity (manifest.json)
//   2. Add `IVX_LIVEKIT` to ScriptingDefineSymbols (Player Settings).

using System;
using System.Threading.Tasks;
using IntelliVerseX.MultiplayerKernel.API;
using UnityEngine;

#if IVX_LIVEKIT
using LiveKit;
using LiveKit.Proto;
#endif

namespace IntelliVerseX.MultiplayerKernel.Voice
{
    /// <summary>
    /// LiveKit-backed implementation of <see cref="IIVXVoice"/>.
    /// Connects to a self-hosted LiveKit SFU using a kernel-minted JWT.
    /// </summary>
    public class IVXLiveKitVoiceProvider : IIVXVoice
    {
        public IVXVoiceProvider Provider => IVXVoiceProvider.LiveKit;

        public IVXVoiceCapability Capability { get; }

        public IVXVoiceMode CurrentMode { get; private set; } = IVXVoiceMode.Off;

        public bool IsConnected { get; private set; }
        public bool IsLocallyMuted { get; private set; }
        public bool HasFloor { get; private set; }

        public event Action<bool> OnConnectionChanged;
        public event Action<IVXSpeakerStateChanged> OnSpeakerStateChanged;
        public event Action<IVXVoiceLevels> OnVoiceLevels;
        public event Action<IVXVoiceMode> OnVoiceModeChanged;
        public event Action<IVXVoiceProvider> OnProviderFailover;
        public event Action<string> OnVoiceUnavailable;

#if IVX_LIVEKIT
        private Room _room;
        private LocalAudioTrack _localAudio;
#endif
        private bool _disposed;

        public IVXLiveKitVoiceProvider(IVXVoiceCapability capability)
        {
            Capability = capability ?? IVXVoiceCapability_Defaults();
        }

        private static IVXVoiceCapability IVXVoiceCapability_Defaults() => new IVXVoiceCapability
        {
            CanPublish = true,
            CanSubscribe = true,
            CanSpatial = true,
            Codecs = new[] { IVXVoiceCodec.Opus },
            MaxPublishers = 16,
            CanChangeProvider = true,
            CanPassthroughExternal = false,
            PttSupported = true,
            BroadcastSupported = true,
            SpatialSupported = true,
        };

        public async Task ConnectAsync(IVXVoiceSessionToken token)
        {
            if (token == null) throw new ArgumentNullException(nameof(token));
            if (token.Provider != IVXVoiceProvider.LiveKit)
            {
                Debug.LogWarning($"[IVXLiveKitVoiceProvider] token provider mismatch: {token.Provider}");
            }
            if (string.IsNullOrEmpty(token.Token) || string.IsNullOrEmpty(token.Url))
            {
                OnVoiceUnavailable?.Invoke("livekit_token_missing");
                return;
            }

#if !IVX_LIVEKIT
            Debug.LogWarning("[IVXLiveKitVoiceProvider] IVX_LIVEKIT scripting define is OFF — voice degrade to none.");
            OnVoiceUnavailable?.Invoke("livekit_sdk_not_compiled");
            await Task.CompletedTask;
            return;
#else
            _room = new Room();
            _room.ParticipantConnected += OnParticipantConnected;
            _room.ParticipantDisconnected += OnParticipantDisconnected;
            _room.ConnectionStateChanged += OnLkConnectionStateChanged;
            _room.ActiveSpeakersChanged += OnLkActiveSpeakersChanged;

            var connectOpts = new RoomOptions
            {
                AutoSubscribe = token.CanSubscribe,
                AdaptiveStream = true,
                Dynacast = true,
            };
            await _room.Connect(token.Url, token.Token, connectOpts);
            IsConnected = true;
            OnConnectionChanged?.Invoke(true);

            if (token.CanPublish)
            {
                _localAudio = await LocalAudioTrack.CreateAudioTrackAsync("ivx-mic");
                await _room.LocalParticipant.PublishAudioTrack(_localAudio);
            }
#endif
        }

        public async Task DisconnectAsync()
        {
#if IVX_LIVEKIT
            if (_localAudio != null) { _localAudio.Dispose(); _localAudio = null; }
            if (_room != null)
            {
                await _room.Disconnect();
                _room.Dispose();
                _room = null;
            }
#endif
            IsConnected = false;
            OnConnectionChanged?.Invoke(false);
            await Task.CompletedTask;
        }

        public async Task SetLocalMuteAsync(bool muted)
        {
            IsLocallyMuted = muted;
#if IVX_LIVEKIT
            if (_localAudio != null) await _localAudio.SetMute(muted);
#endif
            await Task.CompletedTask;
        }

        public Task RequestSpeakerAsync(string topicHint = null)
        {
            // The kernel is the floor authority. The adapter signals the
            // request over the IVX wire (OP_CONV_SPEAKER_REQUEST) — this
            // method is a no-op at the LiveKit layer; HasFloor flips when
            // OnSpeakerStateChanged arrives.
            return Task.CompletedTask;
        }

        public Task ReleaseSpeakerAsync()
        {
            return Task.CompletedTask;
        }

        public async Task PublishSpatialPositionAsync(IVXPoseFrameRef frameRef, float x, float y, float z, float yawDeg)
        {
            // Spatial audio in LiveKit is driven via DataPacket metadata
            // routed through the SFU; the actual positional rendering
            // happens client-side via WebAudio panners. We forward the
            // pose as a small data packet here.
#if IVX_LIVEKIT
            if (_room == null || _room.LocalParticipant == null) return;
            var bytes = System.Text.Encoding.UTF8.GetBytes(
                $"{{\"frame\":\"{frameRef?.FrameId ?? string.Empty}\",\"x\":{x:F3},\"y\":{y:F3},\"z\":{z:F3},\"yaw\":{yawDeg:F1}}}");
            await _room.LocalParticipant.PublishData(bytes, DataPacketKind.Lossy);
#else
            await Task.CompletedTask;
#endif
        }

        public Task SetVoiceModeAsync(IVXVoiceMode mode)
        {
            CurrentMode = mode;
            OnVoiceModeChanged?.Invoke(mode);
            return Task.CompletedTask;
        }

        // ── LiveKit event hooks ───────────────────────────────────────
#if IVX_LIVEKIT
        private void OnParticipantConnected(RemoteParticipant p)            { /* tracked via kernel PLAYER_JOINED */ }
        private void OnParticipantDisconnected(RemoteParticipant p)         { /* tracked via kernel PLAYER_LEFT  */ }
        private void OnLkConnectionStateChanged(ConnectionState state)
        {
            IsConnected = state == ConnectionState.ConnConnected;
            OnConnectionChanged?.Invoke(IsConnected);
            if (state == ConnectionState.ConnDisconnected)
            {
                OnVoiceUnavailable?.Invoke("livekit_disconnected");
            }
        }
        private void OnLkActiveSpeakersChanged(System.Collections.Generic.IReadOnlyList<Participant> speakers)
        {
            // Translate LK active-speakers into IVX VoiceLevels. The kernel
            // also emits a server-side broadcast at 4 Hz; this hook gives
            // local UI a low-latency speaker indicator between broadcasts.
            var samples = new IVXVoiceLevels.Sample[speakers.Count];
            for (int i = 0; i < speakers.Count; i++)
            {
                samples[i] = new IVXVoiceLevels.Sample
                {
                    UserId = speakers[i].Identity,
                    TalkingPct = (uint)Mathf.RoundToInt((speakers[i].AudioLevel) * 100f),
                    Silent = false,
                };
            }
            OnVoiceLevels?.Invoke(new IVXVoiceLevels { Samples = samples, TsMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() });
        }
#endif

        // ── Kernel-driven hooks (called by IVXMatchSession on inbound) ──
        public void OnKernelSpeakerStateChanged(IVXSpeakerStateChanged ev)
        {
            HasFloor = ev != null && ev.Granted;
            OnSpeakerStateChanged?.Invoke(ev);
        }

        public void OnKernelProviderFailover(IVXVoiceProvider next)
        {
            OnProviderFailover?.Invoke(next);
            _ = DisconnectAsync();
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _ = DisconnectAsync();
        }
    }
}

File: SDKs/visionos/Sources/IVXMultiplayer/Voice/IVXLiveKitVoiceProvider.swift
// IVXLiveKitVoiceProvider — visionOS / iOS / macOS implementation of
// IVXVoiceProviderProtocol against client-sdk-swift (LiveKit).
//
// On visionOS the renderer attaches incoming audio tracks to RealityKit
// SpatialAudioComponent for true positional audio. The kernel's
// SpatialFrame translates remote pose to RealityKit anchor space at the
// adapter boundary (see Spatial/IVXSpatialFrameRealityKit.swift).

import Foundation
import LiveKit

@MainActor
public final class IVXLiveKitVoiceProvider: IVXVoiceProviderProtocol, @unchecked Sendable {

    public let provider: IVXVoiceProvider = .livekit
    public let capability: IVXVoiceCapability
    public private(set) var currentMode: IVXVoiceMode = .off
    public private(set) var isConnected: Bool = false
    public private(set) var isLocallyMuted: Bool = false
    public private(set) var hasFloor: Bool = false

    public var onConnectionChanged:    (@Sendable (Bool) -> Void)?
    public var onSpeakerStateChanged:  (@Sendable (IVXSpeakerStateChanged) -> Void)?
    public var onVoiceLevels:          (@Sendable (IVXVoiceLevels) -> Void)?
    public var onVoiceModeChanged:     (@Sendable (IVXVoiceMode) -> Void)?
    public var onProviderFailover:     (@Sendable (IVXVoiceProvider) -> Void)?
    public var onVoiceUnavailable:     (@Sendable (String) -> Void)?

    private var room: Room?

    public init(capability: IVXVoiceCapability = .ivxCanonical()) {
        self.capability = capability
    }

    public func connect(token: IVXVoiceSessionToken) async throws {
        guard !token.token.isEmpty, !token.url.isEmpty else {
            onVoiceUnavailable?("livekit_token_missing")
            return
        }

        let room = Room()
        self.room = room
        room.add(delegate: self)

        let connectOptions = ConnectOptions(autoSubscribe: token.canSubscribe)
        let roomOptions = RoomOptions(
            adaptiveStream: true,
            dynacast: true
        )
        do {
            try await room.connect(url: token.url, token: token.token, connectOptions: connectOptions, roomOptions: roomOptions)
        } catch {
            onVoiceUnavailable?("livekit_connect_failed: \(error)")
            return
        }

        isConnected = true
        onConnectionChanged?(true)

        if token.canPublish {
            do {
                try await room.localParticipant.setMicrophone(enabled: true)
            } catch {
                onVoiceUnavailable?("livekit_mic_failed: \(error)")
            }
        }
    }

    public func disconnect() async {
        if let room = room {
            await room.disconnect()
            self.room = nil
        }
        isConnected = false
        onConnectionChanged?(false)
    }

    public func setLocalMute(_ muted: Bool) async {
        isLocallyMuted = muted
        guard let room = room else { return }
        do {
            try await room.localParticipant.setMicrophone(enabled: !muted)
        } catch {
            // Mute failures are surfaced via OnVoiceUnavailable so the
            // game UI can prompt for OS permissions if needed.
            onVoiceUnavailable?("livekit_mute_failed: \(error)")
        }
    }

    public func requestSpeaker(topicHint: String?) async {
        // Floor authority is in the kernel; this is a no-op at the SFU layer.
    }

    public func releaseSpeaker() async {
        // No-op at SFU layer; kernel revokes via SpeakerStateChanged.
    }

    public func publishSpatialPosition(frameRef: IVXPoseFrameRef, x: Float, y: Float, z: Float, yawDeg: Float) async {
        guard let room = room else { return }
        let payload: [String: Any] = [
            "frame": frameRef.frameId,
            "x": x, "y": y, "z": z,
            "yaw": yawDeg, "ts": frameRef.tsMs
        ]
        guard let bytes = try? JSONSerialization.data(withJSONObject: payload) else { return }
        try? await room.localParticipant.publish(data: bytes, options: DataPublishOptions(reliability: .lossy))
    }

    public func setVoiceMode(_ mode: IVXVoiceMode) async {
        currentMode = mode
        onVoiceModeChanged?(mode)
    }

    // ---- Kernel-driven hooks ----
    public func onKernelSpeakerStateChanged(_ ev: IVXSpeakerStateChanged) {
        hasFloor = ev.granted
        onSpeakerStateChanged?(ev)
    }

    public func onKernelProviderFailover(_ next: IVXVoiceProvider) {
        onProviderFailover?(next)
        Task { await disconnect() }
    }
}

extension IVXLiveKitVoiceProvider: RoomDelegate {
    nonisolated public func room(_ room: Room, didUpdateConnectionState newState: ConnectionState, from oldState: ConnectionState) {
        Task { @MainActor in
            self.isConnected = newState == .connected
            self.onConnectionChanged?(self.isConnected)
            if newState == .disconnected {
                self.onVoiceUnavailable?("livekit_disconnected")
            }
        }
    }

    nonisolated public func room(_ room: Room, didUpdateSpeakers speakers: [Participant]) {
        let samples = speakers.map { p in
            IVXVoiceLevelsSample(
                userId: p.identity?.stringValue ?? "",
                talkingPct: UInt32((p.audioLevel * 100).rounded()),
                silent: false
            )
        }
        let levels = IVXVoiceLevels(samples: samples, tsMs: Int64(Date().timeIntervalSince1970 * 1000))
        Task { @MainActor in self.onVoiceLevels?(levels) }
    }
}
